File: other/day02.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func compiler(program []int) (result int) {
	i := 0

	for true {

		opcode := program[i]
		//a := int(opcode/100000) % 10
		//b := int(opcode/10000) % 10
		//c := int(opcode/1000) % 10
		opcode = opcode % 100

		if opcode == 99 {
			result = program[0]
			return
		}

		value1 := program[program[i+1]]
		value2 := program[program[i+2]]
		value3 := program[i+3]

		if opcode == 1 {
			program[value3] = value1 + value2
			i += 4
		} else if opcode == 2 {
			program[value3] = value1 * value2
			i += 4
		}
	}
	result = program[0]
	return
}

func part1(program []int) (result int) {

	program[1] = 12
	program[2] = 2

	result = compiler(program)
	return
}

func part2(program []int) (answer int) {
	expected := 19690720

	for noun := 80; noun < 100; noun++ {
		for verb := 80; verb < 100; verb++ {

			newProgram := program

			newProgram[1] = noun
			newProgram[2] = verb

			result := compiler(newProgram)
			fmt.Println(result)
			if result == expected {
				answer = noun*100 + verb
				return
			}
		}
	}
	return
}

func main() {
	filename := "./input/day2input.txt"
	dat, err := ioutil.ReadFile(filename)
	check((err))

	arr1 := strings.Split(string(dat), ",")
	var arr2 = []int{}

	for _, element := range arr1 {
		num, err := strconv.Atoi(element)
		check(err)
		arr2 = append(arr2, num)
	}

	program1 := arr2
	result1 := part1(program1)
	fmt.Println("Part 1 = ", result1)

	program2 := arr2
	result2 := part2(program2)
	fmt.Println("Part 2 = ", result2)
}

File: 2025/day02.py
def part1(content, value = 0):
    
    for line in content:
        a, b = map(int, line.split('-'))

        for id in range(a, b + 1):
            id = str(id)
            middle = len(id) // 2
            if id[:middle] == id[middle:]:
                value += int(id)

    print(value)

def part2(content, value = 0):
    
    for line in content:
        a, b = map(int, line.split('-'))

        for id in range(a, b + 1):
            id = str(id)

            for length in range(1, (len(id)//2 + 1)):
                for index in range(0, len(id), length):
                    if id[index: index + length] != id[:length]:
                        break
                else:
                    value += int(id)
                    break

    print(value)

if __name__ == "__main__":
    with open('input/day02.txt') as f:
        content = f.read().split(',')
        part1(content) # 31210613313
        part2(content) # 41823587546
